src/client/game.cpp

```diff
@@ -2479,6 +2479,18 @@ void Game::updatePlayerControl(const CameraOrientation &cam)
 		keypress_bits |= 1U << 4;
 	}
 
+	// autoforward if set: simulate "up" key
+	if (player->getPlayerSettings().continuous_forward) {
+		// simulate "aux1" key if "up" key is already set
+		if (!control.up) {
+			control.up = true;
+			keypress_bits |= 1U << 0;
+		} else {
+			control.aux1 = true;
+			keypress_bits |= 1U << 5;
```

Comment:
I don't think aux1 can be simulated when 'up' is pressed while continuous forward is enabled. Because aux1/special/special1 is an auxilary control that can be set to control anything, it isn't always 'Move fast in fast mode'.
We also have 'aux1_descends = false' that makes aux1 control descending.

Reply:
Ah, I see.
However, this is part of the fix. Before, it was to possible to walk without pressing up and to superspeed without pressing aux1 and up (only one was needed). And such a behaviour can be considered as buggy because there are many things that rely on (1) up (or down, …) and not aux1 being pressed if a player walks normally and (2) up (or down, …) and aux1 being pressed if a player superspeeds. Examples for such things are sprint mods or the thing that makes the player animation faster if aux1 is pressed.
Silent moving forward and silent superspeeding is what this PR should fix.
You could argue the same way by saying that there are other things that up is used for than just for moving up. For example the minetest_game boat moves forward if you press up, ergo moving forward is something else than pressing up.
If someone wants to superspeed in a climbable node, they should disable 'aux1_descends' and not have to use continuous_forward to do some buggy thing.

Reply:
> there are many things that rely on [...] (2) up (or down, …) and aux1 being pressed if a player superspeeds

Yes, unfortunately, but there's no avoiding this.
It's not just about the 'aux1_descends' setting, aux1 can be used to control any operation, so it cannot be automatically 'pressed' when 'up' and 'autoforward' are both active.

As a key, the player can choose to not press aux1, but with your PR as it is, aux1 is always activated in a certain situation and there's no avoiding this. Any mod that uses aux1 for something other than superspeed could not be used.
Setting aux1 as pressed is very different from setting 'superspeed' to true.
Perhaps this is what nerzhul meant by "aux1 is a key there is no key to be pressed for superspeed?".

This could possibly be resolved later by making 'fast' a separate key from 'aux1', as has already been suggested.

I would like input from other core devs, but as far as i can see what is needed is:
Autoforward sends 'up' keypresses to the server, but not 'aux1' keypresses.
The `if (control.up) {` code block in `applyControl()` should not be changed.
All else in the PR looks ok.

Reply:
I too would not prefer to automatically set `aux1` when autoforward is enabled, for the above reasons.

Reply:
Giving players the ability to superspeed without pressing aux1 is the same as giving them the ability to walk forward without pressing up.

>As a key, the player can choose to not press aux1, but with your PR as it is, aux1 is always activated in a certain situation and there's no avoiding this.

It's only activated if the players presses up while continuous_forward is enabled. The player can choose to not superspeed while moving continuously forward, which means that aux1 is not set.

The up key is useless with continuous_forward enabled. This is why it activates aux1 (for example in order to superspeed).

Normally a player doesn't hold up while he uses autoforward (the reason for autoforward is that you do not need to press up).

Reply:
> Normally a player doesn't hold up while he uses autoforward (the reason for autoforward is that you do not need to press up).

That's because `up` is dedicated to moving forward, be it the player, or the entity they're attached to. This is obviously not the case for `aux1`, which is a "super key" for implementing anything and everything that makes use of a keypress as trigger. Sprint is only one (albeit a popular one) use-case. While some very specific actions can be bound to many other keys via the "Change Keys" dialog, only `aux1` is sent to the server, which is the sole reason so many mods use this key.

The real problem is that the engine has hard-coded superspeed to the `aux1` key, which is very unfortunate, in hindsight. As paramat noted, we could (read should) separate superspeed as a separate key.

Reply:
Would you be happy if `up` did not set `aux1` with `continuous_forward` but let's say `down` or `right`? Or should I maybe just remove the feature that you can superspeed with `continuous_forward` and `up` entirely?
Superspeeding is one use of `aux1` and `aux1` is also used for other things, but it should not be possible to superspeed without pressing `aux1` as much as it should not be possible to move forward without pressing `up`.

Reply:
> Or should I maybe just remove the feature that you can superspeed with continuous_forward and up entirely?

That's the way to go, IMHO. Don't take my word for it, though. We'll wait for core dev opinions.

I also agree that superspeed should not be activated without setting any other key.

Reply:
> It's only activated if the players presses up while continuous_forward is enabled.

I know =)
> The player can choose to not superspeed while moving continuously forward

Yes, but many players will try pressing forward when autoforwarding, and when they do, whatever action has been bound to aux1 will happen, this will be weird, sloppy and buggy behaviour that should not happen. Better to attend to the problems another way, by perhaps creating a dedicated fast key, as was suggested by rubenwardy i think recently.
> Normally a player doesn't hold up while he uses autoforward

Well, obviously =) But they can and will press it, especially since it is currently usable for fast move.
> Would you be happy if up did not set aux1 with continuous_forward but let's say down or right?

No, that is just as bad and has the same issues.
> should I maybe just remove the feature that you can superspeed with continuous_forward and up entirely?

That would be ok with me. However, i'm not completely sure about the best form for this PR, i would like more core dev input.

Reply:
A new superspeed key would be out of scope for this PR.

It would be possible to only activate `aux1` if fast is enabled.

Reply:
> A new superspeed key would be out of scope for this PR.

I agree.

> It would be possible to only activate aux1 if fast is enabled.

Why though? Leaving superspeed out of autoforward is probably the better option, IMO.

Reply:
> > It would be possible to only activate aux1 if fast is enabled.
> 
> Why though? Leaving superspeed out of autoforward is probably the better option, IMO.

It would still be possible to press `aux1` to superspeed.
I'm against removing the feature to superspeed with up in autoforward mode.

Anyway, discussing this midge as if it was an elephant doesn't help to get this PR approved, instead it increases the TL;DR effect and thus increases the time needed for more core dev input which is needed for approval.

Reply:
> A new superspeed key would be out of scope for this PR.

Yes, sorry to be unclear, i meant that work on a dedicated fast key should perhaps be done later.

Autoforward is still not bound to a key by default, and i only made it bindable to a key a little while ago, so i think we're fairly free to alter the feature.
> I'm against removing the feature to superspeed with up in autoforward mode.

Ok, earlier you suggested that, so i thought you supported that. I'm ok with leaving that feature in or removing it.

As i see it, holding 'up' when autoforward is enabled, is very similar to holding 'E', so it seems duplicated. If, as likely, we get a dedicated key for 'fast', it will certainly be duplicated.
I have a preference for removing 'autoforward'+'up' triggering fast, but won't insist.

Reply:
Leaving it as it is will result in all aux1-dependent code getting triggered while pressing up during autoforward, which will result in broken behaviour. While I'd really prefer to remove setting `aux1` when pressing `up`, leaving this as it is is fine too, provided this behaviour is documented in a prominent place.

Reply:
SmallJoker agrees with me, aux should not be simulated.